### PR TITLE
[Geos-11324] MapML WMS Vector Representation Style Classes

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-11324](https://osgeo-org.atlassian.net/browse/GEOS-11324) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Review by prushforth — APPROVED** [review]
LGTM - thanks!

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+import org.geotools.api.style.Rule;
+import org.geotools.api.style.Stroke;
+import org.geotools.api.style.Symbol;
+import org.geotools.api.style.Symbolizer;
+import org.geotools.api.style.TextSymbolizer;
+import org.geotools.filter.visitor.IsStaticExpressionVisitor;
+import org.geotools.styling.AbstractStyleVisitor;
+import org.geotools.util.logging.Logging;
+
+/** MapML Style Visitor to convert SLD to MapML CSS Styles */
```
Comment (aaime) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java:35]: I would expand a bit to describe the intent of the class, e.g. what is supported, and its "best effort" nature.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    public static final String FILL = "fill";
+    public static final String RADIUS = "r";
+    public static final String RULE_SYMBOLIZER_DELIMITER = ":";
+    public static final String STROKE_DASHOFFSET = "stroke-dashoffset";
+    public static final String FILL_OPACITY = "fill-opacity";
+
+    Map<String, MapMLStyle> styles = new HashMap<>();
+    AtomicInteger ruleCounter = new AtomicInteger(0);
+    AtomicInteger symbolizerCounter = new AtomicInteger(0);
+    AtomicBoolean isElseFilter = new AtomicBoolean(false);
```
Comment (aaime) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java:57]: I don't think you need anything atomic here, it's not a concurrent class, and these are fields, so they can be updated without issue... e.g., it should work with simple int and boolean variables?

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            r.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(Fill fill) {
+        if (fill.getGraphicFill() != null) {
+            LOGGER.log(
+                    Level.WARNING,
+                    "MapML feature styling does not currently support Graphic Fills");
```
Comment (aaime) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java:79]: Lots of warnings in this class, but in a running server, that can get old quickly,  if it's receiving a lot of MapML requests the logs will be flooded.
By default, FINE is more than enough. If you want easier debugging, make the Level a static final controllable by a system property.

**Review by aaime — CHANGES_REQUESTED** [review]
Needs some documentation to indicate which  parts of SLD are supported, which not, and what happens when they are not.

Also, while there is testing for the visitor, I see none for the association of features to their rules, and at least one functional test checking that everything is fitting together, and the output contains the right styles and feature <-> style associations.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/xml/Feature.java [hunk 4]
```diff
@@ -80,9 +80,14 @@ public class Feature {
     @XmlSchemaType(name = "anySimpleType")
     protected String id;
 
-    @XmlAttribute(name = "class")
+    @XmlAttribute(name = "featuretype-class")
     @XmlSchemaType(name = "anySimpleType")
     protected String clazz;
```
[the remark this answers is not in the document]
Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/xml/Feature.java:85]: correct

**Comment by turingtestfail** [comment]
> Needs some documentation to indicate which parts of SLD are supported, which not, and what happens when they are not.
> 
> Also, while there is testing for the visitor, I see none for the association of features to their rules, and at least one functional test checking that everything is fitting together, and the output contains the right styles and feature <-> style associations.

Documentation added and functional test added for styles to feature associations

File: src/extension/mapml/src/main/java/org/geoserver/mapml/xml/Feature.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/xml/Feature.java]: done

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java [hunk 5]
```diff
@@ -120,14 +129,77 @@ public static Mapml featureCollectionToMapML(
...
+                if (styles != null) {
+                    List<MapMLStyle> applicableStyles = getApplicableStyles(feature, styles);
+                    Optional<Feature> f =
+                            featureBuilder.buildFeature(
+                                    feature, fCaptionTemplate, applicableStyles);
+                    // feature will be skipped if geometry incompatible with style symbolizer
+                    f.ifPresent(features::add);
+                    // WFS GETFEATURE request with no styles
+                } else {
+                    Optional<Feature> f =
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java]: done

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java [hunk 6]
```diff
@@ -120,14 +129,77 @@ public static Mapml featureCollectionToMapML(
...
+        if (applicableStyles.isEmpty()) {
+            for (MapMLStyle style : styles.values()) {
+                if (style.isElseFilter()) {
+                    applicableStyles.add(style);
+                }
+            }
+        }
+        if (applicableStyles.isEmpty()) {
+            LOGGER.finer("No applicable SLD styles found for feature " + sf.getID());
+        }
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java]: done

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java [hunk 7]
```diff
@@ -120,14 +129,77 @@ public static Mapml featureCollectionToMapML(
...
+            return null;
+        }
+        StringBuilder style = new StringBuilder();
+        for (Map.Entry<String, MapMLStyle> entry : styles.entrySet()) {
+            MapMLStyle mapMLStyle = entry.getValue();
+            // empty properties can happen when style elements are not supported
+            if (mapMLStyle != null && !mapMLStyle.getProperties().isEmpty()) {
+                style.append(STYLE_CLASS_PREFIX);
+                style.append(mapMLStyle.getStyleAsCSS());
+                style.append(STYLE_CLASS_SUFFIX);
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java]: switched to StringJoiner and renamed

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeaturesBuilder.java [hunk 8]
```diff
@@ -104,6 +107,21 @@ public Mapml getMapMLDocument() throws IOException {
...
+            if (styleInfo != null && styleInfo.getStyle() != null) {
+                style = styleInfo.getStyle();
+            } else {
+                throw new ServiceException(
+                        "No style or default style found for layer"
+                                + getMapRequest.getLayers().get(0).getLayerInfo().getName());
+            }
+        }
+        style.accept(styleVisitor);
+        Map<String, MapMLStyle> styles = styleVisitor.getStyles();
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeaturesBuilder.java]: Moved to method

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java [hunk 9]
```diff
@@ -90,8 +109,85 @@ public Feature buildFeature(SimpleFeature sf, String featureCaptionTemplate)
...
+            // convert geometry to type expected by the first symbolizer
+            geometryContent = convertGeometryToSymbolizerType(g, mapMLStyles.get(0));
+            String spaceDelimitedCSSClasses =
+                    mapMLStyles.stream()
+                            .map(MapMLStyle::getCSSClassName)
+                            .collect(Collectors.joining(CSS_STYLE_DELIMITER));
+            f.setStyle(spaceDelimitedCSSClasses);
+        }
+        // can't convert geometry to type expected by symbolizer
+        if (geometryContent.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java:126]: Done

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java (hunk 1 above)
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java]: Done

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java (hunk 2 above)
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java]: Switched to simple int and boolean

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java (hunk 3 above)
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java]: Warnings changed to FINE

**Comment by turingtestfail** [comment]
> Looks like there is a legit test failure:
> 
> ```
> 2024-03-17T15:19:13.0244226Z 15:19:13,024 [INFO] Compiling 18 source files to /home/runner/work/geoserver/geoserver/src/extension/geofence/geofence-server/target/test-classes
> 2024-03-17T15:19:13.3518861Z 15:19:13,351 [ERROR] Tests run: 5, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 1.81 s <<< FAILURE! - in org.geoserver.mapml.MapMLWMSFeatureTest
> 2024-03-17T15:19:13.3524440Z 15:19:13,352 [ERROR] testMapMLUseFeaturesWithSLDFilter(org.geoserver.mapml.MapMLWMSFeatureTest)  Time elapsed: 0.082 s  <<< ERROR!
> 2024-03-17T15:19:13.3526211Z org.geoserver.platform.ServiceException: 
> 2024-03-17T15:19:13.3527293Z javax.xml.bind.UnmarshalException
> 2024-03-17T15:19:13.3530476Z  - with linked exception:
> 2024-03-17T15:19:13.3532332Z [Exception [EclipseLink-25008] (Eclipse Persistence Services - 2.7.4.v20190115-ad5b7c6b2a): org.eclipse.persistence.exceptions.XMLMarshalException
> 2024-03-17T15:19:13.3534922Z Exception Description: A descriptor with default root element {http://www.opengis.net/ogc}ServiceExceptionReport was not found in the project]
> 2024-03-17T15:19:13.3537177Z 	at org.geoserver.mapml.MapMLWMSFeatureTest.testMapMLUseFeaturesWithSLDFilter(MapMLWMSFeatureTest.java:139)
> 2024-03-17T15:19:13.3538746Z Caused by: javax.xml.bind.UnmarshalException
> 2024-03-17T15:19:13.3540315Z 	at org.geoserver.mapml.MapMLWMSFeatureTest.testMapMLUseFeaturesWithSLDFilter(MapMLWMSFeatureTest.java:139)
> 2024-03-17T15:19:13.3542216Z Caused by: org.eclipse.persistence.exceptions.XMLMarshalException: 
> 2024-03-17T15:19:13.3543135Z 
> 2024-03-17T15:19:13.3544323Z Exception Description: A descriptor with default root element {http://www.opengis.net/ogc}ServiceExceptionReport was not found in the project
> 2024-03-17T15:19:13.3546479Z 	at org.geoserver.mapml.MapMLWMSFeatureTest.testMapMLUseFeaturesWithSLDFilter(MapMLWMSFeatureTest.java:139)
> ```

Not sure, but I think something is wrong with the Linux test host?  I fixed those test errors several commits ago.  The fix showed up in the Mac and Windows tests but for some reason I can't get the remote Linux tests to pass despite it passing on my local Linux instance.  One of the failures is a missing file that I can confirm is present??

**Comment by prushforth** [comment]
I get the same error locally:

```log
Tests run: 5, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 1.742 s <<< FAILURE! - in org.geoserver.mapml.MapMLWMSFeatureTest
testMapMLUseFeaturesWithSLDFilter(org.geoserver.mapml.MapMLWMSFeatureTest)  Time elapsed: 0.049 s  <<< ERROR!
org.geoserver.platform.ServiceException: 
javax.xml.bind.UnmarshalException
 - with linked exception:
[Exception [EclipseLink-25008] (Eclipse Persistence Services - 2.7.4.v20190115-ad5b7c6b2a): org.eclipse.persistence.exceptions.XMLMarshalException
Exception Description: A descriptor with default root element {http://www.opengis.net/ogc}ServiceExceptionReport was not found in the project]
	at org.geoserver.mapml.MapMLWMSFeatureTest.testMapMLUseFeaturesWithSLDFilter(MapMLWMSFeatureTest.java:139)
Caused by: javax.xml.bind.UnmarshalException
	at org.geoserver.mapml.MapMLWMSFeatureTest.testMapMLUseFeaturesWithSLDFilter(MapMLWMSFeatureTest.java:139)
Caused by: org.eclipse.persistence.exceptions.XMLMarshalException: 

Exception Description: A descriptor with default root element {http://www.opengis.net/ogc}ServiceExceptionReport was not found in the project
	at org.geoserver.mapml.MapMLWMSFeatureTest.testMapMLUseFeaturesWithSLDFilter(MapMLWMSFeatureTest.java:139)

testMapMLGetStyleQuery(org.geoserver.mapml.MapMLWMSFeatureTest)  Time elapsed: 0.011 s  <<< ERROR!
java.io.IOException: No such resource: polygonOneFilter.sld
	at org.geoserver.mapml.MapMLWMSFeatureTest.testMapMLGetStyleQuery(MapMLWMSFeatureTest.java:210)

```

**Comment by prushforth** [comment]
Pretty odd failure - tests are green on windows, but on ubuntu I can actually see the file it's complaining about in the test's classes directory, but in the debugger, catalog.getStyleByName("polygonFilter").getSLD() reports that it can't find the polygonOneFilter.sld file.

**Comment by turingtestfail** [comment]
> Pretty odd failure - tests are green on windows, but on ubuntu I can actually see the file it's complaining about in the test's classes directory, but in the debugger, catalog.getStyleByName("polygonFilter").getSLD() reports that it can't find the polygonOneFilter.sld file.

Found it.  It was a case sensitivity issue with the file name.  Fix pushed.

**Comment by prushforth** [comment]
> Found it. It was a case sensitivity issue with the file name. Fix pushed.

Good eye!

**Review by aaime — APPROVED** [review]

**Comment by aaime** [comment]
Merged on main. Not backported on 2.25.x as we are in Release Candidate mode (no new features in core/extension between RC and final). We can backport once 2.25.0 is out.
